### PR TITLE
Added Loyalty System (10.74)

### PR DESCRIPTION
**Pull request by reyaleman**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
* Implemented [LOYALTY SYSTEM](https://tibia.fandom.com/wiki/Loyalty_System)
* System is optional, can be disabled in config.lua (loyaltyEnabled = false)
* Loyalty bonuses are customizables in config.lua
* Modified Player::GetPercentLevel to return floats (client support decimal numbers on skill bars)
* Loyalty Titles onLook
* Loyalty message bonus on serverlog (Player:onLogin script)
* Kept skill & tries untouched, loyalty is calculated when the player is loaded.
* Loyalty points are updated on startup but only if it's been more than 1 day since last update (server config)
* Added 3 lua functions (Player:getLoyaltyStatus, Player:addLoyalty & Player:removeLoyalty)
* Added migration file to update database (column "loyalty_points" added to accounts table and inserted server config "loyalty_update" value to update loyalty on a daily basis)

#### Missing:
* Social loyalty badges (social not implemented yet)
* Loyalty highscores (highscores not implemented yet)

loyalty

**Issues addressed:** 
#3871 Version 10.74 (The Loyalty System was introduced.)

Any comments would be highly appreciated. 😃 

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -362,7 +363,7 @@ CREATE TABLE IF NOT EXISTS `towns` (
   UNIQUE KEY `name` (`name`)
 ) ENGINE=InnoDB DEFAULT CHARACTER SET=utf8;
 
-INSERT INTO `server_config` (`config`, `value`) VALUES ('db_version', '32'), ('players_record', '0');
+INSERT INTO `server_config` (`config`, `value`) VALUES ('db_version', '32'), ('players_record', '0'), ('loyalty_updated', '0');
```
Comment (DSpeichert) [inline schema.sql:366]:
```suggestion
INSERT INTO `server_config` (`config`, `value`) VALUES ('db_version', '33'), ('players_record', '0'), ('loyalty_updated', '0');
```

I think you forgot to bump the schema version here.

**Review by DSpeichert — CHANGES_REQUESTED** [review]

File: schema.sql (hunk 1 above)
Comment (reyaleman) [reply schema.sql:366]: thanks! I'll fix it in the next commit, I'm still waiting more reviews

File: src/player.cpp [hunk 2]
```diff
@@ -4835,3 +4960,39 @@ void Player::updateRegeneration()
...
+	}
+
+	loyaltyPoints += points;
+	return true;
+}
+
+std::string Player::getLoyaltyTitleDescription() const
+{
+	if (loyaltyPoints <= 50) {
+		return "a Scout of Tibia";
```
Comment (omarcopires) [inline src/player.cpp:4977]: maybe we should use world name or citizen instead of tibia

Comment (reyaleman) [reply src/player.cpp:4977]: okay, I'll use ConfigManager::SERVER_NAME

File: src/player.cpp [hunk 3]
```diff
@@ -564,6 +579,65 @@ void Player::removeSkillTries(skills_t skill, uint64_t count, bool notify /* = f
...
+}
+
+void Player::updateLoyaltySkill(uint8_t skill)
+{
+	if (skill > SKILL_LAST) {
+		return;
+	}
+
+	skills[skill].bLevel = skills[skill].level;
+	skills[skill].bTries = skills[skill].tries;
```
Comment (nekiro) [inline src/player.cpp:632]: is bLevel supposed to be baseLevel?
If so what do you think about just naming it baseLevel, baseTries and so on

File: src/player.cpp [hunk 4]
```diff
@@ -1778,6 +1853,57 @@ void Player::removeManaSpent(uint64_t amount, bool notify /* = false*/)
...
+		currReqManaSpent = nextReqManaSpent;
+		nextReqManaSpent = vocation->getReqMana(bMagLevel + 1);
+		if (currReqManaSpent >= nextReqManaSpent) {
+			count = 0;
+			break;
+		}
+	}
+
+	magLevelPercent = Player::getPercentLevel(count, nextReqManaSpent);
+	bManaSpent += count;
```
Comment (nekiro) [inline src/player.cpp:1880]: same here, baseManaSpent?

File: src/player.cpp [hunk 5]
```diff
@@ -4835,3 +4961,39 @@ void Player::updateRegeneration()
...
+{
+	if (loyaltyPoints == 0) {
+		return false;
+	}
+
+	loyaltyPoints += points;
+	return true;
+}
+
+std::string Player::getLoyaltyTitleDescription() const
```
Comment (nekiro) [inline src/player.cpp:4975]: this should be a "tool" method, so place this in tools.cpp and pass loyaltyPoints, you can also put everything into a map and just iterate it to find proper title (code will be shorter and easier to edit)

File: src/player.cpp [hunk 6]
```diff
@@ -4835,3 +4961,39 @@ void Player::updateRegeneration()
 		condition->setParam(CONDITION_PARAM_MANATICKS, vocation->getManaGainTicks() * 1000);
 	}
 }
+
+bool Player::addLoyalty(int32_t points)
+{
+	if (loyaltyPoints == 0) {
```
Comment (nekiro) [inline src/player.cpp:4967]: I might not understand how this work, but why would you not allow adding loyalty points if loyaltyPoints is 0?

**Review by nekiro — CHANGES_REQUESTED** [review]
Great stuff

File: src/player.cpp (hunk 6 above)
Comment (nekiro) [reply src/player.cpp:4967]: Not sure, but I think it doesnt matter, so just let it work even with loyalty system disabled

File: src/configmanager.cpp [hunk 7]
```diff
@@ -334,6 +364,19 @@ float ConfigManager::getExperienceStage(uint32_t level) const
 	return std::get<2>(*it);
 }
 
+float ConfigManager::getLoyaltyBonus(uint16_t points) const
+{
+	auto it = std::find_if(loyaltyBonuses.begin(), loyaltyBonuses.end(), [points](LoyaltyBonuses::value_type bonus) {
```
Comment (nekiro) [inline src/configmanager.cpp:369]: `auto it = std::find_if(loyaltyBonuses.begin(), loyaltyBonuses.end(), [points](const auto& bonus) {`?

File: src/luascript.cpp [hunk 8]
```diff
@@ -10800,6 +10805,49 @@ int LuaScriptInterface::luaPlayerGetIdleTime(lua_State* L)
...
+{
+	// player:removeLoyalty([points = 1])
+	Player* player = getUserdata<Player>(L, 1);
+	if (!player) {
+		lua_pushnil(L);
+	}
+
+	uint16_t points = getNumber<uint16_t>(L, 2, 1);
+	pushBoolean(L, player->addLoyalty(-points));
+
```
Comment (nekiro) [inline src/luascript.cpp:10847]: remove newline

File: src/luascript.cpp [hunk 9]
```diff
@@ -10800,6 +10805,49 @@ int LuaScriptInterface::luaPlayerGetIdleTime(lua_State* L)
...
+{
+	// player:addLoyalty([points = 1])
+	Player* player = getUserdata<Player>(L, 1);
+	if (!player) {
+		lua_pushnil(L);
+	}
+
+	uint16_t points = getNumber<uint16_t>(L, 2, 1);
+	pushBoolean(L, player->addLoyalty(points));
+
```
Comment (nekiro) [inline src/luascript.cpp:10833]: remove newline

File: src/player.cpp [hunk 10]
```diff
@@ -146,6 +151,16 @@ std::string Player::getDescription(int32_t lookDistance) const
 		} else {
 			s << " has no vocation.";
 		}
+
+		if (g_config.getBoolean(ConfigManager::LOYALTY_SYSTEM) && loyaltyPoints >= 50) {
+			if (sex == PLAYERSEX_FEMALE) {
```
Comment (nekiro) [inline src/player.cpp:156]: use fmt::format?

File: src/player.cpp [hunk 11]
```diff
@@ -126,6 +126,11 @@ std::string Player::getDescription(int32_t lookDistance) const
 		} else {
 			s << " You have no vocation.";
 		}
+
+		if (g_config.getBoolean(ConfigManager::LOYALTY_SYSTEM) && loyaltyPoints >= 50) {
+			s << " You are " << getLoyaltyTitleDescription(loyaltyPoints) << " "
```
Comment (nekiro) [inline src/player.cpp:131]: use fmt::format?

File: src/player.cpp [hunk 12]
```diff
@@ -1778,6 +1870,72 @@ void Player::removeManaSpent(uint64_t amount, bool notify /* = false*/)
...
+	if (!isLogin) {
+		float bonusCount = count * (1 + loyaltyBonus);
+
+		float amount, fractional;
+		fractional = std::modf(bonusCount, &amount);
+		manaSpentFractional += std::round(fractional * 100) / 100;
+
+		count = static_cast<uint64_t>(amount);
+
+		if (manaSpentFractional == (int)manaSpentFractional) {
```
Comment (nekiro) [inline src/player.cpp:1929]: use static_cast and int32_t

**Comment by omarcopires** [comment]
Please don't leave this!!

File: config.lua.dist [hunk 13]
```diff
@@ -135,6 +135,22 @@ experienceStages = {
 	{ minlevel = 101, multiplier = 3 }
 }
 
+-- Loyalty system
+loyaltyEnabled = true
```
Comment (ranisalt) [inline config.lua.dist:139]: You can drop this flag and let users effectively disable the loyalty system by removing all tiers.

**Comment by omarcopires** [comment]
could you leave some extra functions in LUA to be used in cyclopedia?
`getLoyaltyMagicLevel`, `getLoyaltySkill` and `getLoyaltyTitleDescription`

I also think it would be better to move the descriptions to onLook instead of keeping them directly in the engine

File: data/globalevents/scripts/startup.lua [hunk 14]
```diff
@@ -42,4 +42,18 @@ function onStartup()
 		local position = town:getTemplePosition()
 		db.query("INSERT INTO `towns` (`id`, `name`, `posx`, `posy`, `posz`) VALUES (" .. town:getId() .. ", " .. db.escapeString(town:getName()) .. ", " .. position.x .. ", " .. position.y .. ", " .. position.z .. ")")
 	end
+	
+	--  loyalty system (update points)
+	if configManager.getBoolean(configKeys.LOYALTY_SYSTEM) then
+		local updatedAt = 0
+		local resultId = db.storeQuery("SELECT `value` FROM `server_config` WHERE `config` = 'loyalty_updated'")
```
Comment (omarcopires) [inline data/globalevents/scripts/startup.lua:49]:
```lua
local resultId = ...
if not resultId then
  return {}
end
```

**Comment by reyaleman** [comment]
I'd have to rewrite this and make a new pull request (taking your comments in consideration), IIRC I stopped this because I was having issues with float values being inexact 👀

**Pull request closed without merging**